### PR TITLE
Linkify

### LINKED ISSUE #46: Allow method chaining without rebuilding the tree
Right now, if you chain methods (e.g. `linkify(clean(text))`) the whole tree-building, which is expensive, has to happen twice.

I can see two ways of solving this. One is a total API rewrite, one is less complicated. I'm leaning toward the latter.

**Total rewrite:**

Recreate a `Bleach` class, but this time it _is_ the text you want to clean, e.g.:

``` python
from bleach import Bleach
text = Bleach(text)
print text.clean()
```

It could store state of what's already been done (like cleaning, linkifying, etc) and either return text (as above) or return itself to allow chaining, e.g.:

``` python
# Implement __unicode__ and __str__, seems nicer than above.
print Bleach(text).clean().linkify()
```

**Less-crazy idea:**

Instead of actual `unicode` objects, the top-level methods return a lazy object that _looks_ like a string, maybe even descends from `unicode`, but is smarter, stores the tree, and only renders itself if anything needs to treat it as a string.

Then the methods would be reworked to check for the tree before rebuilding it, and if found, they would just use the existing tree. `clean()` might need special handling--i.e. always go back to text first--because of the nature of what it does. But I generally say you should always call `clean()` first if you're chaining, anyway.

**Pros and cons**

The rewrite might make it easier to deal with the increasing complexity of `linkify()`. You could (again) subclass `Bleach` and override a number of things there. I don't love APIs where you have to subclass things. I'm also not sure how complex this would make the upgrade for consumers--it would be a major version change.

The less-crazy idea doesn't involve subclassing anything, ever. The API can stay the same and it can be a minor version change. I am guessing won't be _too_ difficult to write some sort of `LazyTreeStringObjectThing` class. It would be a good opportunity to do some refactoring and get some stuff out of `bleach/__init__.py`. (I wouldn't change import paths, just where the code lives.)

### LINKED ISSUE #130: bleach linkify unescapes escaped HTML sometimes
@jsocol, thanks for your library, we use it for years :+1: 

Recently we discovered quite a strange behaviour `linkify` exposes in one scenario. Please, take a look on my example below. As you can see, sometimes it unescapes a correctly escaped tag:

``` python

In [1]: import bleach

In [2]: bleach.__version__
Out[2]: u'1.4'

In [3]: bleach.linkify('&lt;br&gt;')  # everything is okay
Out[3]: u'&lt;br&gt;'

In [4]: bleach.linkify('&lt;br&gt; http://example.com')  # why <br> is unescaped?
Out[4]: u'<br> <a href="http://example.com" rel="nofollow">http://example.com</a>'

In [5]: bleach.linkify('&lt;br&gt; <br> http://example.com')  # ... and here it's not again ;-)
Out[5]: u'&lt;br&gt; <br> <a href="http://example.com" rel="nofollow">http://example.com</a>'
```

Frankly, I have no idea how it should be fixed. Thanks for any input.

### LINKED ISSUE #200: linkify hanging for this string:
With fresh install of bleach (1.4.3) in a new virtualenv:

```
>>> import bleach
>>> bleach.linkify("an@email.com<mailto:an@email.com>", parse_email=True)
... hangs ...
```

(found because I have it linkifying emails - and this came through in someone's sig)

### PR DESCRIPTION
**Pull request by willkg**
This is a rewrite of linkify and all that stuff.

Fixes #46.

Fixes #200.

Fixes #130.

### CODE REVIEW

File: bleach/linkifier.py [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+                    in_a = True
+                    token_buffer.append(token)
+
+                    # We buffer the start tag, so we don't want to yield it,
+                    # yet
+                    continue
+
+            elif in_pre:
+                # NOTE(willkg): We put this clause here since in_a and
+                # switching in and out of is_a takes precedence.
```
Comment (willkg) [inline bleach/linkifier.py]: Typo--the `is_a` should be an `in_a`.

File: bleach/utils.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+
+def alphabetize_attributes(attrs):
+    """Takes a dict of attributes (or None) and returns them alphabetized"""
+    if not attrs:
+        return attrs
+
+    return OrderedDict(
+        [(k, v) for k, v in sorted(attrs.items(), key=_attr_key)]
+    )
```
Comment (willkg) [inline bleach/utils.py]: I should add some tests for this.

**Comment by willkg** [comment]
This allows people to do use cases specified in #181, so I think this covers that, too.

**Comment by willkg** [comment]
This also covers #165.

**Comment by willkg** [comment]
This covers #153.

**Comment by willkg** [comment]
This covers #78, too.

File: bleach/linkifier.py [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+                        new_tokens.append({u'type': u'Characters', u'data': text[end:]})
+
+                    for new_token in new_tokens:
+                        yield new_token
+
+                    continue
+
+            yield token
+
+    def strip_non_url_bits(self, fragment):
```
Comment (willkg) [inline bleach/linkifier.py]: One nice thing about pulling all this functionality into this single method is that someone can override/extend it with a `SuperLinkifyDoWhatIWantFilter`.

**Comment by willkg** [comment]
Seems good to me. Landing this!

**Comment by kousu** [comment]
I really appreciate you taking a look at #130 and getting this out to PyPI! From my perspective, it was fixed almost immediately. That's amazing! :clap:

**Pull request merged**